doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1588,39 +1588,70 @@ Controlling the proof flow
 
 .. tacn:: specialize @one_term_with_bindings {? @as_ipat }
 
-   This tactic works on the :term:`head constant` of the :n:`@one_term` in
-   :n:`@one_term_with_bindings`.
-   The premises of this hypothesis (either universal quantifications or
-   :term:`non-dependent premises <non-dependent premise>`) are instantiated by
-   concrete terms from :n:`@one_term_with_bindings`.
-
-   .. todo PR:
-      1) what is "non-dependent implication" above?
-      2) In this little example, what exactly is (H x)? It looks almost like
-         a function but H doesn't look like a function.  Is this just something
-         wierd about specialize or have I missed some basic point?
-
-      Goal forall x: nat, (forall n, n >= 0) -> True.
-      intros.  (* H is forall n : nat, n >= 0 *)
-      specialize (H x). (* H is x >= 0 *)
-
-   The application to :n:`@one_term` can be partial. The
-   uninstantiated arguments are inferred by unification if possible or left
-   quantified in the hypothesis otherwise. With the :n:`as` clause, the local
-   hypothesis :n:`@ident` is left unchanged and instead, the modified hypothesis
-   is introduced as specified by the :token:`simple_intropattern`. The name :n:`@ident`
-   can also refer to a global lemma or hypothesis. In this case, for
-   compatibility reasons, the behavior of :tacn:`specialize` is close to that of
-   :tacn:`generalize`: the instantiated statement becomes an additional premise of
-   the goal. The ``as`` clause is especially useful in this case to immediately
-   introduce the instantiated statement as a local hypothesis.
+   If the :term:`head constant` of :n:`@one_term` has the type
+   `forall ...`, the tactic replaces one or more of the
+   quantified variables in the type with terms provided by
+   :n:`@one_term_with_bindings`, either in the form of a
+   :ref:`function application <function_application>` (which may be partial),
+   such as `(H 1)`, or with named binders, such as `H with (n:=1)`.
+
+   If the :term:`head constant` has a non-dependent product type such as
+   `A -> B -> C`, the tactic does :term:`forward reasoning`, eliminating one or
+   more of the premises.
```

Comment:
In the dependent case it is also forward reasoning: replacing `H:forall a b:A, (T a b)` by `forall b:A, T 1 b` is also eliminating a premise. I think you can just remove that sentence.

Reply:
The doc describes forward and backward reasoning [here](https://coq.gitlab.io/-/coq/-/jobs/3454025112/artifacts/_build/default/doc/refman-html/proof-engine/tactics.html#term-forward-reasoning).  The sense for "forward reasoning" described [here](https://en.wikipedia.org/wiki/Forward_chaining) makes sense for `A -> B -> C` but I don't see how it makes sense for dependent products.

Also the second example refers to an `apply` example using forward reasoning in hypotheses, so it made sense to mention the term here.